drgn_tools/iscsi.py

```diff
@@ -0,0 +1,213 @@
+# Copyright (c) 2024, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Helper for iscsi
+"""
+import argparse
+from typing import Iterator
+from typing import Tuple
+
+from drgn import container_of
+from drgn import Object
+from drgn import Program
+from drgn import sizeof
+from drgn.helpers.common import escape_ascii_string
+from drgn.helpers.common.type import enum_type_to_class
+from drgn.helpers.linux.list import list_for_each_entry
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.scsi import for_each_scsi_host
+from drgn_tools.scsi import host_module_name
+from drgn_tools.table import print_dictionary
+from drgn_tools.table import print_table
+from drgn_tools.util import enum_name_get
+from drgn_tools.util import has_member
+
+ISCSI_SESSION_STATES = ["LOGGED_IN", "FAILED", "FREE"]
+
+
+def for_each_iscsi_tcp_host(prog: Program) -> Iterator[Object]:
+    """
+    Iterates through all scsi host devices and returns a
+    iterator of hosts of iscsi_tcp type
+    :returns: a iterator of ``struct Scsi_Host *``
+    """
+    for shost in for_each_scsi_host(prog):
+        if host_module_name(shost) == "iscsi_tcp":
+            yield shost
+
+
+def __get_iscsi_sw_tcp_host(prog: Program, iscsi_tcp_host: Object) -> Object:
+    """
+    Get the iscsi_sw_tcp_host asscociated with an iscsi tcp host
+    :returns: ``struct iscsi_sw_tcp_host``
+    """
+    return Object(
+        prog,
+        "struct iscsi_sw_tcp_host",
+        address=iscsi_tcp_host.hostdata.address_of_().value_()
+        + sizeof(prog.type("struct iscsi_host")),
+    )
+
+
+def shost_for_each_device(prog: Program, shost: Object) -> Iterator[Object]:
+    """
+    Get a list of scsi_devices asscociated with an Scsi_Host
+    :returns: a iterator of ``struct scsi_device``
+    """
+    return list_for_each_entry(
+        "struct scsi_device", shost.__devices.address_of_(), "siblings"
+    )
+
+
+def get_dev_name(prog: Program, sdev: Object) -> str:
+    """
+    Get the device name associated with scsi_device.
+    :return ``str``
+    """
+    rq = sdev.request_queue
+    dev = container_of(rq.kobj.parent, "struct device", "kobj")
+    return dev.kobj.name.string_().decode()
+
+
+def for_each_iscsi_session(prog: Program) -> Iterator[Object]:
+    """
+    Iterates through all iscsi tcp hosts, gets their associated iscsi_sw_tcp_hosts and return a
+    iterator of their sessions.
+    :returns: a iterator of ``struct iscsi_session *``
+
+    """
+    for h in for_each_iscsi_tcp_host(prog):
+        yield __get_iscsi_sw_tcp_host(prog, h).session
+
```

Comment:
As in my following comment, let try to make these helper generic to iscsi drivers, not only to iscsi_tcp driver.